### PR TITLE
fix: allow empty slowStart when using LeastRequest

### PR DESCRIPTION
**Pull request by zirain**
fixes: #4202

### CODE REVIEW

**Comment by arkodg** [comment]
@zirain can you help explain how this fix is related to the issue
from the issue, I see this error
```
"error": "loadBalancer setting is invalid, only one setting can be set"}
```
which points to this https://github.com/envoyproxy/gateway/blob/75fa107d3d985ced40dc777a9118909938cfe230/internal/ir/xds.go#L1908

**Comment by zirain** [comment]
> loadBalancer setting is invalid

if you use `LeastRequest` with empty `SlowStart`, leads lb.LeastRequest to empty, that cause `Validate()` failed with 0 matched.

```golang
	case egv1a1.LeastRequestLoadBalancerType:
		lb = &ir.LoadBalancer{}
		if policy.LoadBalancer.SlowStart != nil {
			if policy.LoadBalancer.SlowStart.Window != nil {
				lb.LeastRequest = &ir.LeastRequest{
					SlowStart: &ir.SlowStart{
						Window: policy.LoadBalancer.SlowStart.Window,
					},
				}
			}
		}
```

File: internal/gatewayapi/clustersettings.go [hunk 1]
```diff
@@ -299,7 +299,11 @@ func buildLoadBalancer(policy egv1a1.ClusterSettings) (*ir.LoadBalancer, error)
 			ConsistentHash: consistentHash,
 		}
 	case egv1a1.LeastRequestLoadBalancerType:
-		lb = &ir.LoadBalancer{}
+		lb = &ir.LoadBalancer{
```
Comment (arkodg) [inline internal/gatewayapi/clustersettings.go:302]: should this be ?
```
	case egv1a1.LeastRequestLoadBalancerType:
		lb = &ir.LoadBalancer{
			LeastRequest: &ir.LeastRequest{},
		}
		if policy.LoadBalancer.SlowStart != nil {
			if policy.LoadBalancer.SlowStart.Window != nil {
				lb.LeastRequest.SlowStart = &ir.SlowStart{
				    Window: policy.LoadBalancer.SlowStart.Window,
				}
			}
		}
```		

and same for `RoundRobin` ?

Comment (zirain) [reply internal/gatewayapi/clustersettings.go:302]: I reviewed the logic, this's better.

File: internal/gatewayapi/clustersettings.go [hunk 2]
```diff
@@ -315,17 +315,13 @@ func buildLoadBalancer(policy egv1a1.ClusterSettings) (*ir.LoadBalancer, error)
...
 		}
-		if policy.LoadBalancer.SlowStart != nil {
-			if policy.LoadBalancer.SlowStart.Window != nil {
-				lb.RoundRobin = &ir.RoundRobin{
-					SlowStart: &ir.SlowStart{
-						Window: policy.LoadBalancer.SlowStart.Window,
-					},
-				}
+		if policy.LoadBalancer.SlowStart != nil && policy.LoadBalancer.SlowStart.Window != nil {
+			lb.RoundRobin = &ir.RoundRobin{
```
Comment (arkodg) [inline internal/gatewayapi/clustersettings.go:321]: `lb.RoundRobin.SlowStart` = .....

File: internal/gatewayapi/clustersettings.go [hunk 3]
```diff
@@ -299,14 +299,14 @@ func buildLoadBalancer(policy egv1a1.ClusterSettings) (*ir.LoadBalancer, error)
...
-				lb.LeastRequest = &ir.LeastRequest{
-					SlowStart: &ir.SlowStart{
-						Window: policy.LoadBalancer.SlowStart.Window,
-					},
-				}
+		lb = &ir.LoadBalancer{
+			LeastRequest: &ir.LeastRequest{},
+		}
+		if policy.LoadBalancer.SlowStart != nil && policy.LoadBalancer.SlowStart.Window != nil {
+			lb.LeastRequest = &ir.LeastRequest{
```
Comment (arkodg) [inline internal/gatewayapi/clustersettings.go:306]: lb.LeastRequest.SlowStart = .....

File: internal/xds/translator/cluster.go [hunk 4]
```diff
@@ -183,15 +183,13 @@ func buildXdsCluster(args *xdsClusterArgs) *clusterv3.Cluster {
...
 	} else if args.loadBalancer.RoundRobin != nil {
 		cluster.LbPolicy = clusterv3.Cluster_ROUND_ROBIN
-		if args.loadBalancer.RoundRobin.SlowStart != nil {
-			if args.loadBalancer.RoundRobin.SlowStart.Window != nil {
-				cluster.LbConfig = &clusterv3.Cluster_RoundRobinLbConfig_{
-					RoundRobinLbConfig: &clusterv3.Cluster_RoundRobinLbConfig{
-						SlowStartConfig: &clusterv3.Cluster_SlowStartConfig{
-							SlowStartWindow: durationpb.New(args.loadBalancer.RoundRobin.SlowStart.Window.Duration),
-						},
+		if args.loadBalancer.RoundRobin.SlowStart != nil && args.loadBalancer.RoundRobin.SlowStart.Window != nil {
```
Comment (arkodg) [inline internal/xds/translator/cluster.go:186]: just indentation change, or did anything change here ?

Comment (zirain) [reply internal/xds/translator/cluster.go:186]: Reduce cyclomatic complexity?

**Review by arkodg — APPROVED** [review]
thanks !

**Review by guydc — APPROVED** [review]

**Pull request merged**